### PR TITLE
Fix stuck query after cancel or termination when segment is not responding

### PR DESCRIPTION
**Pull request by whitehawk**
Fix stuck query after cancel or termination when segment is not responding (#1132)

Problem:
The following scenario caused a stuck state of a coordinator backend process:
1. At some moment one of the segments stopped processing incoming requests (the
reason itself is not important, as sometimes bad things may happen with any
segment, and the system should be able to recover).
2. A query, which dispatches requests to segments, was executed (for example,
select from 'gp_toolkit.gp_resgroup_status_per_segment'). As one of the segments
was not responding, the coordinator hanged in 'checkDispatchResult' (it is
expected, and can be handled by the FTS).
3. The stuck query was canceled or terminated (by 'pg_cancel_backend' or
'pg_terminate_backend'). It didn't help to return from the stuck state, but
after this step the query became completely unrecoverable. Even after FTS had
detected the malfunction segment and had promoted the mirror, the query was
still hanging forever.

Expected behavior is that after FTS mirror promotion, all stuck queries are
unblocked and canceled successfully.

Note: if FTS mirror promotion happened before step 3, the FTS canceled the
query successfully.

Root cause:
During cancel or termination, the coordinator tried to do abort of the current
transaction, and it hanged in the function 'internal_cancel' (called from
PQcancel) on 'poll' system call. It has a timeout of 660 seconds, and, moreover,
after the timeout expired, it looped forever trying to do 'poll' again. So, if
the socket was opened on the segment side, but nobody replied (as the segment
process became not available for some reason), 'internal_cancel' had no way to
return.

Fix:
Once FTS promotes a mirror, it sends the  signal to the coordinator
postmaster (with PMSIGNAL_FTS_PROMOTED_MIRROR reason). On receiving of the
signal, the coordinator postmaster sends the SIGUSR1 signal (with
PROCSIG_FTS_PROMOTED_MIRROR reason) to all of its usual backends. Once the
backend receives the signal, if it is in the state of cancelling or terminating
of the query, it sets a flag in libpq. 'internal_cancel' checks this flag before
calling the 'poll' system call. If it is set, it will return with an error.
Thus:
a. if the FTS promotion happens before cancel/terminate, the query will be
canceled by the old logic;
b. if the FTS promotion happens after cancel/terminate, but before the
'internal_cancel' calls the 'poll', 'internal_cancel' will return an error
without calling the 'poll';
c. if the FTS promotion happens when the 'poll' is already called, the 'poll'
will return EINTR (as the SIGUSR1 was received), and a new 'poll' will not be
called, as the flag is set.

(cherry picked from commit https://github.com/arenadata/gpdb/commit/96ed1ad4825f593b02c24c9426273894da18b3ba)

Differences comparing to the original commit:
1. exec_cmd_on_segments() is placed inside setup.sql file instead of
server_helper.sql, as the latter doesn't exist in GPDB 7 (and all functions from
this file are located in setup.sql)
2. The check 'GpIdentity.segindex == MASTER_CONTENT_ID' in 'sigusr1_handler()'
is replaced with 'IS_QUERY_DISPATCHER()', which is equivalent. It is done,
because MASTER_CONTENT_ID was renamed to COORDINATOR_CONTENT_ID in GPDB 7.
3. 'TermSignalReceived' global var was removed in GPDB 7 (when it was merged
with PostgreSQL 9.5), so it is brought back.
4. exec_cmd_on_segments() is reworked due to limitation in GPDB 7. Previous
version failed with "ERROR:  EXECUTE ON ALL SEGMENTS is only supported for
set-returning functions", so now it returns a table with TEXT field instead of
a single string. Also, exec_cmd_on_segments() is changed to use 'plpython3u'
language instead of 'plpythonu'.
5. Test is updated to work on GPDB 7.
6. File 'isolation2_resgroup_schedule' doesn't exist anymore, so the test is
added to 'isolation2_resgroup_v1_schedule' and 'isolation2_resgroup_v2_schedule'
files.

-----------------------

Should be rebased.

### CODE REVIEW

**Comment by RekGRpth** [comment]
> Fix stuck query after cancel or termination when segment is not responding (#948)

How about to change this to

```
Fix stuck query after cancel or termination when segment is not responding (#1132)
```

File: src/backend/utils/init/globals.c [hunk 1]
```diff
@@ -40,6 +40,8 @@ volatile sig_atomic_t QueryCancelCleanup = false;
 volatile sig_atomic_t QueryCancelPending = false;
 volatile sig_atomic_t QueryFinishPending = false;
 
+volatile bool TermSignalReceived = false;
```
Comment (RekGRpth) [inline src/backend/utils/init/globals.c]:
```suggestion
volatile sig_atomic_t TermSignalReceived = false;
```

**Comment by RekGRpth** [comment]
> 6\. isolation2_resgroup_v2_schedule

Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!

**Comment by whitehawk** [comment]
> > Fix stuck query after cancel or termination when segment is not responding (#948)
> 
> How about to change this to
> 
> ```
> Fix stuck query after cancel or termination when segment is not responding (#1132)
> ```

I'm Ok to update it.

But that's a general issue when we do cherry-picking. Please consider adding this point about updating PR reference to our checklist.

**Comment by whitehawk** [comment]
> > 6. isolation2_resgroup_v2_schedule
> 
> Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!

There is a note in the v1 schedule file:
```
# This file should be consistent with isolation2_resgroup_v2_schedule,
# if you add/delete test file, as well do it at isolation2_resgroup_v2_schedule.
```

**Comment by RekGRpth** [comment]
> Please consider adding this point about updating PR reference to our checklist.

Yes, it would be good if everyone did this to avoid confusion.

**Comment by RekGRpth** [comment]
> > > 6. isolation2_resgroup_v2_schedule
> > 
> > 
> > Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!
> 
> There is a note in the v1 schedule file:
> 
> ```
> # This file should be consistent with isolation2_resgroup_v2_schedule,
> # if you add/delete test file, as well do it at isolation2_resgroup_v2_schedule.
> ```

I was not asking about the need to add a test for the second version, but about checking this test with the second version.

**Comment by whitehawk** [comment]
> > Fix stuck query after cancel or termination when segment is not responding (#948)
> 
> How about to change this to
> 
> ```
> Fix stuck query after cancel or termination when segment is not responding (#1132)
> ```

Updated

**Comment by whitehawk** [comment]
> > > > 6. isolation2_resgroup_v2_schedule
> > > 
> > > 
> > > Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!
> > 
> > 
> > There is a note in the v1 schedule file:
> > ```
> > # This file should be consistent with isolation2_resgroup_v2_schedule,
> > # if you add/delete test file, as well do it at isolation2_resgroup_v2_schedule.
> > ```
> 
> I was not asking about the need to add a test for the second version, but about checking this test with the second version.

Checked locally, the test passes:
```
test setup                        ... ok         2412 ms
test resgroup/resgroup_auxiliary_tools_v2 ... ok        20405 ms (diff   57 ms)
test resgroup/resgroup_views_seg_down ... ok        74791 ms (diff   62 ms)
```

**Comment by RekGRpth** [comment]
> > > > > 6. isolation2_resgroup_v2_schedule
> > > > 
> > > > 
> > > > Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!
> > > 
> > > 
> > > There is a note in the v1 schedule file:
> > > ```
> > > # This file should be consistent with isolation2_resgroup_v2_schedule,
> > > # if you add/delete test file, as well do it at isolation2_resgroup_v2_schedule.
> > > ```
> > 
> > 
> > I was not asking about the need to add a test for the second version, but about checking this test with the second version.
> 
> Checked locally, the test passes:
> 
> ```
> test setup                        ... ok         2412 ms
> test resgroup/resgroup_auxiliary_tools_v2 ... ok        20405 ms (diff   57 ms)
> test resgroup/resgroup_views_seg_down ... ok        74791 ms (diff   62 ms)
> ```

The test should not only pass, but also fail without a patch.

**Comment by whitehawk** [comment]
> > > > > > 6. isolation2_resgroup_v2_schedule
> > > > > 
> > > > > 
> > > > > Is it necessary to check how the test works with resource groups of the second version? This has not been checked on CI yet!
> > > > 
> > > > 
> > > > There is a note in the v1 schedule file:
> > > > ```
> > > > # This file should be consistent with isolation2_resgroup_v2_schedule,
> > > > # if you add/delete test file, as well do it at isolation2_resgroup_v2_schedule.
> > > > ```
> > > 
> > > 
> > > I was not asking about the need to add a test for the second version, but about checking this test with the second version.
> > 
> > 
> > Checked locally, the test passes:
> > ```
> > test setup                        ... ok         2412 ms
> > test resgroup/resgroup_auxiliary_tools_v2 ... ok        20405 ms (diff   57 ms)
> > test resgroup/resgroup_views_seg_down ... ok        74791 ms (diff   62 ms)
> > ```
> 
> The test should not only pass, but also fail without a patch.

It does fail when the fix is disabled, like in sample patch below:

```diff
diff --git a/src/interfaces/libpq/fe-connect.c b/src/interfaces/libpq/fe-connect.c
index 19ed3597de..198c548574 100644
--- a/src/interfaces/libpq/fe-connect.c
+++ b/src/interfaces/libpq/fe-connect.c
@@ -4378,7 +4378,7 @@ PQfreeCancel(PGcancel *cancel)
  */
 void PQbypassConnCloseAtCancel(pqbool bypass)
 {
-       bypass_conn_close_at_cancel = bypass;
+       //bypass_conn_close_at_cancel = bypass;
 }
 
 /*

```

it results in expected fail:

```diff
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /usr/re/share/ws87-ADBDEV-5704/gpdb/src/test/isolation2/expected/resgroup/resgroup_views_seg_down.out /usr/re/share/ws87-ADBDEV-5704/gpdb/src/test/isolation2/results/resgroup/resgroup_views_seg_down.out
--- /usr/re/share/ws87-ADBDEV-5704/gpdb/src/test/isolation2/expected/resgroup/resgroup_views_seg_down.out	2024-11-28 17:33:45.425425296 +1000
+++ /usr/re/share/ws87-ADBDEV-5704/gpdb/src/test/isolation2/results/resgroup/resgroup_views_seg_down.out	2024-11-28 17:33:45.434425297 +1000
@@ -87,7 +104,7 @@
 SELECT count(*) = 0 AS query_complete FROM pg_stat_activity WHERE state = 'active' AND query = 'SELECT count(1) FROM gp_toolkit.gp_resgroup_status_per_segment WHERE groupname = ''admin_group'';';
  query_complete 
 ------------
- t              
+ f              
 (1 row)
 
 -- Recover back the segment
@@ -216,7 +577,7 @@
 SELECT count(*) = 0 AS query_complete FROM pg_stat_activity WHERE state = 'active' AND query = 'SELECT count(1) FROM gp_toolkit.gp_resgroup_status_per_segment WHERE groupname = ''admin_group'';';
  query_complete 
 ------------
- t              
+ f              
 (1 row)
 
 -- Recover back the segment

```

**Review by RekGRpth — APPROVED** [review]

**Review by bimboterminator1 — APPROVED** [review]

**Pull request merged**